dangerzone/document.py

```diff
@@ -12,29 +12,57 @@ class DocumentHolder(object):
     Keeps track of one document
     """
 
-    def __init__(self, input_file_path: str) -> None:
-        # Name of input and out files
-        self._input_filename: Optional[str] = input_file_path
+    def __init__(self, input_file_path: str = None) -> None:
+        self._input_filename: Optional[str] = None
         self._output_filename: Optional[str] = None
 
+        if input_file_path:
+            self.input_filename = input_file_path
+
     @property
     def input_filename(self) -> str:
         if self._input_filename is None:
-            raise RuntimeError("Input filename has not been set yet.")
+            raise DocumentFilenameException("Input filename has not been set yet.")
         else:
             return self._input_filename
 
     @input_filename.setter
     def input_filename(self, filename: str) -> None:
+        # validate input filename
+        try:
+            open(filename, "rb")
+        except FileNotFoundError:
+            raise DocumentFilenameException(
+                "Input file not found: make sure you typed it correctly."
+            )
+        except PermissionError:
+            raise DocumentFilenameException(
+                "You don't have permission to open the input file."
+            )
+
         self._input_filename = filename
 
     @property
     def output_filename(self) -> str:
         if self._output_filename is None:
-            raise RuntimeError("Output filename has not been set yet.")
+            raise DocumentFilenameException("Output filename has not been set yet.")
         else:
             return self._output_filename
 
     @output_filename.setter
     def output_filename(self, filename: str) -> None:
+        # validate output filename
+        if not filename.endswith(".pdf"):
+            raise DocumentFilenameException("Safe PDF filename must end in '.pdf'")
+        try:
+            with open(os.path.abspath(filename), "wb"):
+                pass
+        except PermissionError:
+            raise DocumentFilenameException("Safe PDF filename is not writable")
+
         self._output_filename = filename
+
+
+class DocumentFilenameException(Exception):
+    def __init__(self, message):
+        super().__init__(message)
```

Comment:
Also, the `DocumentFilenameException()` class is introduced in 'simplify "new window" logic", and is patched in "add unit tests for document.py". It's best to keep thing clean and merge the patch in the commit that introduces this class.

Reply:
> Also, the DocumentFilenameException() class is introduced in 'simplify "new window" logic", and is patched in "add unit tests for document.py". 

Now it's introduced in `validate I/O filenames in Document class` and is no longer patched there.

> I'm not sure about the `DocumentFilenameException` as is,

This a UX thing. Front ends (closer to the user) should present more actionable information. Otherwise, on every front end we'll have to catch the original exception and show the user-understandable translation. But using an intermediary exception
  * Instead of showing the error: `[Errno 2] No such file or directory: document.pdf` which is not user-actionable
  * instead we show: `Input file not found: make sure you typed it correctly.`


> because we lose traceback info and the original message. We could retain these with raise ... from ..., but is it worth the extra complexity?

I have added the `raise ... from ...` syntax. I was not aware this was a thing.

Reply:
Cool, thanks for the changes. As for the original traceback, now that we retain it with `raise ... from ...`, it would be nice logging it as well. This requires handling the `DocumentFilenameException` from a single place. I'll give it a try.